util/dbutil/database.go

```diff
@@ -186,54 +194,87 @@ type Config struct {
 	ConnMaxLifetime string `yaml:"conn_max_lifetime"`
 }
 
+type Config struct {
+	PoolConfig   `yaml:",inline"`
+	ReadOnlyPool PoolConfig `yaml:"ro_pool"`
+}
+
 func (db *Database) Configure(cfg Config) error {
-	db.RawDB.SetMaxOpenConns(cfg.MaxOpenConns)
-	db.RawDB.SetMaxIdleConns(cfg.MaxIdleConns)
+	if err := db.configure(db.ReadOnlyDB, cfg.ReadOnlyPool); err != nil {
+		return err
+	}
+
+	return db.configure(db.RawDB, cfg.PoolConfig)
+}
+
+func (db *Database) configure(rawDB *sql.DB, cfg PoolConfig) error {
+	if rawDB == nil {
+		return nil
+	}
+
+	rawDB.SetMaxOpenConns(cfg.MaxOpenConns)
+	rawDB.SetMaxIdleConns(cfg.MaxIdleConns)
 	if len(cfg.ConnMaxIdleTime) > 0 {
 		maxIdleTimeDuration, err := time.ParseDuration(cfg.ConnMaxIdleTime)
 		if err != nil {
 			return fmt.Errorf("failed to parse max_conn_idle_time: %w", err)
 		}
-		db.RawDB.SetConnMaxIdleTime(maxIdleTimeDuration)
+		rawDB.SetConnMaxIdleTime(maxIdleTimeDuration)
 	}
 	if len(cfg.ConnMaxLifetime) > 0 {
 		maxLifetimeDuration, err := time.ParseDuration(cfg.ConnMaxLifetime)
 		if err != nil {
 			return fmt.Errorf("failed to parse max_conn_idle_time: %w", err)
 		}
-		db.RawDB.SetConnMaxLifetime(maxLifetimeDuration)
+		rawDB.SetConnMaxLifetime(maxLifetimeDuration)
 	}
 	return nil
 }
 
 func NewFromConfig(owner string, cfg Config, logger DatabaseLogger) (*Database, error) {
-	dialect, err := ParseDialect(cfg.Type)
-	if err != nil {
-		return nil, err
-	}
-	conn, err := sql.Open(cfg.Type, cfg.URI)
+	wrappedDB, err := NewWithDialect(cfg.URI, cfg.Type)
 	if err != nil {
 		return nil, err
 	}
 	if logger == nil {
 		logger = NoopLogger
 	}
-	wrappedDB := &Database{
-		RawDB: conn,
 
-		Owner:   owner,
-		Dialect: dialect,
+	wrappedDB.Owner = owner
+	wrappedDB.Log = logger
 
-		Log: logger,
+	if cfg.ReadOnlyPool.MaxOpenConns > 0 {
+		if cfg.ReadOnlyPool.Type == "" {
+			cfg.ReadOnlyPool.Type = cfg.Type
+		}
+		if cfg.ReadOnlyPool.URI == "" {
+			cfg.ReadOnlyPool.URI = cfg.URI
+		}
 
-		IgnoreForeignTables: true,
-		VersionTable:        "version",
+		uriParts := strings.Split(cfg.ReadOnlyPool.URI, "?")
+
+		var qs url.Values
+		if len(uriParts) == 2 {
+			var err error
+			qs, err = url.ParseQuery(uriParts[1])
+			if err != nil {
+				return nil, err
+			}
+
+			qs.Del("_txlock")
+		}
+		qs.Set("_query_only", "true")
```

Comment:
The uri modification should probably only happen if the read-only pool uri is empty